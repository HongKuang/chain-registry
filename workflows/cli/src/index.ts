export const cli = 1;