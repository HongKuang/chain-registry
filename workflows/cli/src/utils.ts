import chalk from 'chalk';

import { readAndParsePackageJson } from "./package";

// Function to display the version information
export function displayVersion(): void {
  const pkg = readAndParsePackageJson();
  console.log(chalk.green(`Name: ${pkg.name}`));
  console.log(chalk.blue(`Version: ${pkg.version}`));
}

export function getVersion(): string {
  const pkg = readAndParsePackageJson();
  return pkg.version;
}

// MARKED AS NOT DRY
export function toCamelCase(key: string) {
  return key
      // First, remove all leading non-alphabet characters except $
      .replace(/^[^a-zA-Z$]+/, '')
      // Convert what follows a separator into upper case
      .replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '')
      // Ensure the first character of the result is always lowercase
      .replace(/^./, (c) => c.toLowerCase());
}
