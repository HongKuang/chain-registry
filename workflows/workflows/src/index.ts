export * from './registry';
export * from './schema-typescript';
export { JSONSchema } from 'schema-typescript';
