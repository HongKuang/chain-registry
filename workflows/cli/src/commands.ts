import { AssetList, Chain,IBCInfo } from '@chain-registry/types'
import { assetLists, chains, FilePathInfo, ibcInfo, JSONSchema,registryDir, schemas } from '@chain-registry/workflows';
import { existsSync, readFileSync } from 'fs';
import { CLIOptions, Inquirerer } from 'inquirerer'
import { ParsedArgs } from 'minimist';
import { join, resolve } from 'path';

interface RegistryPaths {
  chains: ResolvedPath[];
  assetLists: ResolvedPath[];
  ibcInfo: ResolvedPath[];
  schemas: ResolvedPath[];
}

interface Registry {
  chains: {
    filepath: string;
    content: Chain
  }[];
  assetLists: {
    filepath: string;
    content: AssetList
  }[];
  ibcInfo: {
    filepath: string;
    content: IBCInfo
  }[];
  schemas: {
    filepath: string;
    content: JSONSchema
  }[];
}


export const commands = async (argv: Partial<ParsedArgs>, prompter: Inquirerer, _options: CLIOptions) => {
  let command = argv._[0] as string; // Safely assume command is a string

  if (!command) {
    ({ command } = await prompter.prompt(argv, [
      {
        type: 'autocomplete',
        name: 'command',
        message: 'choose a command',
        options: [
          'chains',
          'asset-lists',
          'ibc-data',
          'schemas'
        ]
      }
    ]));

  }

  argv = await prompter.prompt(argv, [
    {
      type: 'text',
      name: 'registryDir',
      message: 'provide a registryDir:',
      // default: relative(process.cwd(), registryDir)
      default: registryDir,
      useDefault: true // remove later
    }
  ]);

  if (!existsSync(argv.registryDir)) {
    prompter.close();
    throw new Error('bad registry path!');
  }

  // Resolve paths for each item using the helper function
  const registryPaths: RegistryPaths = {
    chains: resolvePaths(chains, argv.registryDir),
    assetLists: resolvePaths(assetLists, argv.registryDir),
    ibcInfo: resolvePaths(ibcInfo, argv.registryDir),
    schemas: resolvePaths(schemas, argv.registryDir)  // schemas are just strings
  };

  const registry: Registry = parseRegistryFiles(registryPaths);

  switch (command) {
    case 'chains':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chainName',
          message: 'Select a chain:',
          options: registry.chains.map(chain => chain.content.chain_name),
          maxDisplayLines: 15
        }
      ]);
      break;
    case 'asset-lists':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain',
          message: 'Select a chain:',
          options: registry.chains.map(chain => chain.content.chain_name),
          maxDisplayLines: 15
        }
      ]);

      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'asset',
          message: 'Select an asset:',
          options: registry.assetLists
            .filter(list => list.content.chain_name === argv.chainName)
            .flatMap(list=>list.content.assets)
            .map(asset => ({
              name: `${asset.symbol} | ${asset.base}`,
              value: asset
            })),
          maxDisplayLines: 15
        }
      ]);

      console.log(JSON.stringify(argv.asset,null,2))
      break;
    case 'ibc-data':
      console.log(argv);

      if (Array.isArray(argv.chain)) {
        argv['chain-1'] = argv.chain[0];
        argv['chain-2'] = argv.chain[1];
      } else if (typeof argv.chain === 'string') {
        argv['chain-1'] = argv.chain;
      }

      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain-1',
          message: 'Select an IBC chain:',
          options: registry.ibcInfo.map(info => info.content.chain_1 + ' <> ' + info.content.chain_2),
          maxDisplayLines: 5
        }
      ]);
      
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain-2',
          message: 'Select an IBC chain:',
          options: registry.ibcInfo.map(info => info.content.chain_1 + ' <> ' + info.content.chain_2),
          maxDisplayLines: 5
        }
      ]);

      console.log(argv);

      break;
    case 'schemas':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'schemaName',
          message: 'Select a schema:',
          options: registry.schemas.map(schema => schema.content.title),
          maxDisplayLines: 5
        }
      ]);
      break;

    default:
      console.log("No recognized command provided or no command given.");
      break;
  }

  return argv;

};


type ResolvedPath = FilePathInfo & {
  resolvedPath: string;
}

function resolvePaths(items: FilePathInfo[], baseDir: string): ResolvedPath[] {
  return items.map(item => {
    return { ...item, resolvedPath: resolve(join(baseDir, item.filepath)) };
  });
}

function parseRegistryFiles(registryPaths: RegistryPaths): Registry {
  const keys = Object.keys(registryPaths);
  const parsedRegistry: any = {};

  keys.forEach(key => {
    // @ts-ignore
    parsedRegistry[key] = registryPaths[key].map((resolvedInfo: ResolvedPath) => {
      const content = readFileSync(resolvedInfo.resolvedPath, 'utf-8');
      return { filepath: resolvedInfo.resolvedPath, content: JSON.parse(content) };
    });
  });

  return parsedRegistry;
}