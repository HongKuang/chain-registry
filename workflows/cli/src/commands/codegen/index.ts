import { Registry, SchemaTypeGenerator } from '@chain-registry/workflows';
import { existsSync } from 'fs';
import { CLIOptions, Inquirerer } from 'inquirerer';
import { ParsedArgs } from 'minimist';

import { toCamelCase } from '../../utils';

export const commands = async (argv: Partial<ParsedArgs>, prompter: Inquirerer, _options: CLIOptions) => {

  const {
    registryDir,
    outputDir,
    strictTypeSafety,
    camelCase
  } = await prompter.prompt(argv, [
    {
      type: 'text',
      name: 'registryDir',
      message: 'provide a registryDir:'
    },
    {
      type: 'text',
      name: 'outputDir',
      message: 'provide a outputDir:'
    },
    {
      type: 'confirm',
      name: 'strictTypeSafety',
      default: true,
      useDefault: true
    },
    {
      type: 'confirm',
      name: 'camelCase',
      default: true,
      useDefault: false
    }
  ]);

  if (!existsSync(registryDir)) {
    prompter.exit();
    throw new Error('bad registry path!');
  }

  const registry: Registry = new Registry(registryDir);
  const generator = new SchemaTypeGenerator({
    outputDir,
    registry,
    schemaTSOptions: {
      strictTypeSafety,
      camelCase,
      camelCaseFn: toCamelCase
    }
  });

  generator.generateTypes();

  return argv;
};