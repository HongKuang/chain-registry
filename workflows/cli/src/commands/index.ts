import { AssetList, Chain, IBCInfo } from '@chain-registry/types'
import { assetLists, chains, FilePathInfo, ibcInfo, JSONSchema, registryDir, schemas } from '@chain-registry/workflows';
import { existsSync, readFileSync } from 'fs';
import { CLIOptions, Inquirerer, Question } from 'inquirerer'
import { ParsedArgs } from 'minimist';
import { join, relative, resolve } from 'path';

interface RegistryPaths {
  chains: ResolvedPath[];
  assetLists: ResolvedPath[];
  ibcInfo: ResolvedPath[];
  schemas: ResolvedPath[];
}

interface Registry {
  chains: {
    filepath: string;
    content: Chain
  }[];
  assetLists: {
    filepath: string;
    content: AssetList
  }[];
  ibcInfo: {
    filepath: string;
    content: IBCInfo
  }[];
  schemas: {
    filepath: string;
    content: JSONSchema
  }[];
}


export const commands = async (argv: Partial<ParsedArgs>, prompter: Inquirerer, _options: CLIOptions) => {
  let command;

  if (argv._.length > 0) {
    command = argv._[0];
  }

  if (command) {
    argv.command = command;
  }

  const questions: Question[] = [
    {
      type: 'autocomplete',
      name: 'command',
      message: 'choose a command',
      options: [
        'chain',
        'asset-lists',
        'ibc-data',
        'schemas'
      ]
    }
  ];

  ({ command } = await prompter.prompt(argv, questions));

  argv = await prompter.prompt(argv, [
    {
      type: 'text',
      name: 'registryDir',
      message: 'provide a registryDir:',
      default: relative(process.cwd(), registryDir),
      useDefault: true // remove later
    }
  ]);

  if (!existsSync(argv.registryDir)) {
    prompter.exit();
    throw new Error('bad registry path!');
  }

  // Resolve paths for each item using the helper function
  const registryPaths: RegistryPaths = {
    chains: resolvePaths(chains, argv.registryDir),
    assetLists: resolvePaths(assetLists, argv.registryDir),
    ibcInfo: resolvePaths(ibcInfo, argv.registryDir),
    schemas: resolvePaths(schemas, argv.registryDir)  // schemas are just strings
  };

  const registry: Registry = parseRegistryFiles(registryPaths);

  switch (command) {
    case 'chain':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain',
          message: 'Select a chain:',
          options: registry.chains.map(chain => ({ name: chain.content.chain_name, value: chain.content })),
          maxDisplayLines: 15
        }
      ]);
      const chain: Chain = argv.chain;
      console.log(chain.chain_id);
      chain.codebase.cosmos_sdk_version && console.log(chain.codebase.cosmos_sdk_version);
      chain.codebase.cosmwasm_version && console.log(chain.codebase.cosmwasm_version);
      chain.codebase.ibc_go_version && console.log(chain.codebase.ibc_go_version);
      break;
    case 'asset-lists':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain',
          message: 'Select a chain:',
          options: registry.chains.map(chain => chain.content.chain_name),
          maxDisplayLines: 15
        }
      ]);

      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'asset',
          message: 'Select an asset:',
          options: registry.assetLists
            .filter(list => list.content.chain_name === argv.chainName)
            .flatMap(list => list.content.assets)
            .map(asset => ({
              name: `${asset.symbol} | ${asset.base}`,
              value: asset
            })),
          maxDisplayLines: 15
        }
      ]);

      console.log(JSON.stringify(argv.asset, null, 2))
      break;
    case 'ibc-data':
      console.log(argv);

      if (Array.isArray(argv.chain)) {
        argv['chain-1'] = argv.chain[0];
        argv['chain-2'] = argv.chain[1];
      } else if (typeof argv.chain === 'string') {
        argv['chain-1'] = argv.chain;
      }

      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain-1',
          message: 'Select an IBC chain:',
          options: registry.chains.map(chain => chain.content.chain_name),
          maxDisplayLines: 15
        }
      ]);

      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'chain-2',
          message: 'Select an IBC chain:',
          options: registry.chains.map(chain => chain.content.chain_name),
          maxDisplayLines: 15
        }
      ]);

      const chain1 = argv['chain-1'];
      const chain2 = argv['chain-2'];

      // Create an array of the chain names, sort them alphabetically, and then join them with a hyphen
      const sortedChainNames = [chain1, chain2].sort();
      const chainNameString = sortedChainNames.join('-');

      const i = registry.ibcInfo.find(ibc => ibc.filepath.endsWith(chainNameString + '.json'));

      if (i) {
        console.log(i.content.channels);
      }

      break;
    case 'schemas':
      argv = await prompter.prompt(argv, [
        {
          type: 'autocomplete',
          name: 'schemaName',
          message: 'Select a schema:',
          options: registry.schemas.map(schema => null),
          maxDisplayLines: 5
        }
      ]);
      break;

    default:
      console.log(`No recognized command provided or no command given: ${command}`);
      break;
  }

  return argv;

};


type ResolvedPath = FilePathInfo & {
  resolvedPath: string;
}

function resolvePaths(items: FilePathInfo[], baseDir: string): ResolvedPath[] {
  return items.map(item => {
    return { ...item, resolvedPath: resolve(join(baseDir, item.filepath)) };
  });
}

function parseRegistryFiles(registryPaths: RegistryPaths): Registry {
  const keys = Object.keys(registryPaths);
  const parsedRegistry: any = {};

  keys.forEach(key => {
    // @ts-ignore
    parsedRegistry[key] = registryPaths[key].map((resolvedInfo: ResolvedPath) => {
      const content = readFileSync(resolvedInfo.resolvedPath, 'utf-8');
      return { filepath: resolvedInfo.resolvedPath, content: JSON.parse(content) };
    });
  });

  return parsedRegistry;
} 