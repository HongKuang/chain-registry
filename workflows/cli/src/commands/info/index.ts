import { Registry } from '@chain-registry/workflows';
import { existsSync } from 'fs';
import { CLIOptions, Inquirerer, Question } from 'inquirerer'
import { ParsedArgs } from 'minimist';

import { displayAsset } from './asset';
import { displayAssetList } from './asset-list';
import { displayChain } from './chain';

export const commands = async (argv: Partial<ParsedArgs>, prompter: Inquirerer, _options: CLIOptions) => {
  let command;

  if (argv._.length > 0) {
    command = argv._.shift();
  }

  if (command) {
    argv.command = command;
  }

  const questions: Question[] = [
    {
      type: 'autocomplete',
      name: 'command',
      message: 'choose a command',
      options: [
        'chain',
        'asset-list',
        'asset'
      ]
    }
  ];

  ({ command } = await prompter.prompt(argv, questions));

  argv = await prompter.prompt(argv, [
    {
      type: 'text',
      name: 'registryDir',
      message: 'provide a registryDir:'
    }
  ]);

  if (!existsSync(argv.registryDir)) {
    prompter.exit();
    throw new Error('bad registry path!');
  }

  const registry: Registry = new Registry(argv.registryDir);

  switch (command) {
  case 'chain':
    argv = await prompter.prompt(argv, [
      {
        type: 'autocomplete',
        name: 'chain',
        message: 'Select a chain:',
        options: registry.chains.map(chain => ({ name: chain.chain_name, value: chain })),
        maxDisplayLines: 15
      }
    ]);
    displayChain(argv.chain);
    break;

  case 'asset-list':
    argv = await prompter.prompt(argv, [
      {
        type: 'autocomplete',
        name: 'chain',
        message: 'Select a chain:',
        options: registry.chains.map(chain => chain.chain_name),
        maxDisplayLines: 15
      }
    ]);

    const assetList = registry.assetLists.find(a => a.chain_name === argv.chain);
    displayAssetList(assetList);
    break;

  case 'asset':
    argv = await prompter.prompt(argv, [
      {
        type: 'autocomplete',
        name: 'chain',
        message: 'Select a chain:',
        options: registry.chains.map(chain => chain.chain_name),
        maxDisplayLines: 15
      }
    ]);

    argv = await prompter.prompt(argv, [
      {
        type: 'autocomplete',
        name: 'asset',
        message: 'Select an asset:',
        options: registry.assetLists
          .filter(list => list.chain_name === argv.chain)
          .flatMap(list => list.assets)
          .map(asset => ({
            name: asset.base,
            value: asset
          })),
        maxDisplayLines: 15
      }
    ]);

    displayAsset(argv.asset);
    break;

  default:
    console.log(`No recognized command provided or no command given: ${command}`);
    break;
  }

  return argv;

};