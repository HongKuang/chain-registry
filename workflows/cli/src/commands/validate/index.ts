import { Registry, SchemaValidator } from '@chain-registry/workflows';
import { existsSync } from 'fs';
import { CLIOptions, Inquirerer } from 'inquirerer'
import { ParsedArgs } from 'minimist';

export const commands = async (argv: Partial<ParsedArgs>, prompter: Inquirerer, _options: CLIOptions) => {

  argv = await prompter.prompt(argv, [
    {
      type: 'text',
      name: 'registryDir',
      message: 'provide a registryDir:',
      required: true,
    },
    {
      type: 'confirm',
      name: 'useStrict',
      required: true,
      useDefault: true,
      default: false
    },
    {
      type: 'confirm',
      name: 'allErrors',
      required: true,
      useDefault: true,
      default: true
    },
    {
      type: 'text',
      name: 'draft',
      required: true,
      useDefault: true,
      default: '2019-09'
    },
    {
      type: 'confirm',
      name: 'useDefaults',
      required: true,
      useDefault: true,
      default: true
    },
    {
      type: 'text',
      name: 'logLevel',
      required: true,
      useDefault: true,
      default: 'info'
    }
  ]);

  if (!existsSync(argv.registryDir)) {
    prompter.exit();
    throw new Error('bad registry path!');
  }

  const registry = new Registry(argv.registryDir)
  
  const {
    draft,
    allErrors,
    useDefaults,
    useStrict,
    logLevel
  } = argv;

  const validator = new SchemaValidator(registry, {
    draft,
    allErrors,
    useDefaults,
    useStrict,
    logLevel
  });

  validator.validateAllData(true);

  return argv;
};