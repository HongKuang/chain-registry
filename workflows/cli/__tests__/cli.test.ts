import readline from 'readline';
import { Readable, Transform,Writable } from 'stream';
import stripAnsi from 'strip-ansi';

import { CLI, CLIOptions } from '../src/commander';
import { commands } from '../src/commands'

jest.mock('readline');

const snap = (str: any) => expect(str).toMatchSnapshot();

describe('Inquirerer', () => {
  let mockWrite: jest.Mock;
  let mockInput: Readable;
  let mockOutput: Writable;
  let transformStream: Transform;

  let writeResults: string[];
  let transformResults: string[];

  let inputQueue: Array<{ type: 'key' | 'read', value: string }> = [];
  let currentInputIndex: number = 0;

  function setupReadlineMock() {
    readline.createInterface = jest.fn().mockReturnValue({
      question: (questionText: string, cb: (input: string) => void) => {
        // Process the queued inputs when question is called
        const nextInput = inputQueue[currentInputIndex++];
        if (nextInput && nextInput.type === 'read') {
          setTimeout(() => cb(nextInput.value), 350); // Simulate readline delay
        }
      },
      close: jest.fn(),
    });
  }

  function enqueueInputResponse(input: { type: 'key' | 'read', value: string }) {
    if (input.type === 'key') {
  // Push key events directly to mockInput
  // @ts-ignore
      setTimeout(() => mockInput.push(input.value), 350);
    } else {
      // Queue readline responses to be handled by the readline mock
      inputQueue.push(input);
    }
  }

  const KEY_SEQUENCES = {
    ENTER: '\u000d',     // Carriage return
    UP_ARROW: '\u001b[A', // ANSI escape sequence for the up arrow
    DOWN_ARROW: '\u001b[B',
    SPACE: ' '
  };

  beforeEach(() => {
    mockWrite = jest.fn();
    writeResults = [];
    transformResults = [];


    mockInput = new Readable({
      read(size) { }
    });
    // @ts-ignore
    mockInput.setRawMode = jest.fn();  // Mock TTY-specific method if needed

    mockOutput = new Writable({
      write: (chunk, encoding, callback) => {
        const str = chunk.toString();
        writeResults.push(stripAnsi(str));
        mockWrite(str);
        callback();
      }
    });

    // Create the transform stream to log and pass through data
    transformStream = new Transform({
      transform(chunk, encoding, callback) {
        const data = chunk.toString();
        transformResults.push(stripAnsi(data));
        this.push(chunk); // Pass the data through
        callback();
      }
    });

    setupReadlineMock();
    mockInput.pipe(transformStream);
  });

  it('prompts user and correctly processes delayed input', async () => {
    enqueueInputResponse({ type: 'read', value: 'user input' });

    const options: Partial<CLIOptions> = {
        noTty: false,
        input: mockInput,
        output: mockOutput,
        minimistOpts: {
          alias: {
            v: 'version'
          }
        }
      };
    
    const app = new CLI(commands, options, {
        _: ['chains']
    });

    const expectedResult = { username: 'user input' };

    const result = await app.run();

    expect(result).toEqual(expectedResult);
    snap(writeResults);
    snap(transformResults);
  });

});
