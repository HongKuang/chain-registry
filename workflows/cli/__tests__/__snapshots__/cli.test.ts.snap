// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inquirerer prompts user and correctly processes delayed input 1`] = `
{
  "_": [
    "asset-lists",
  ],
  "chain": {
    "$schema": "../chain.schema.json",
    "apis": {
      "grpc": [
        {
          "address": "grpc.8ball.nodestake.top:443",
          "provider": "NodeStake",
        },
        {
          "address": "https://grpc.8ball.nodexcapital.com:443",
          "provider": "NodeX Validator",
        },
      ],
      "rest": [
        {
          "address": "https://rest.8ball.info",
          "provider": "8ball",
        },
        {
          "address": "https://api.8ball.nodestake.top",
          "provider": "NodeStake",
        },
        {
          "address": "https://api-8ball.nodine.id/",
          "provider": "Nodine.ID",
        },
      ],
      "rpc": [
        {
          "address": "https://rpc.8ball.info/",
          "provider": "8ball",
        },
        {
          "address": "https://rpc.8ball.nodestake.top",
          "provider": "NodeStake",
        },
      ],
    },
    "bech32_prefix": "8ball",
    "chain_id": "eightball-1",
    "chain_name": "8ball",
    "codebase": {
      "binaries": {
        "linux/amd64": "https://8ball.info/8ball.tar.gz",
      },
      "compatible_versions": [
        "v1",
      ],
      "cosmos_sdk_version": "0.46.7",
      "cosmwasm_enabled": true,
      "genesis": {
        "genesis_url": "https://8ball.info/8ball-genesis.json",
      },
      "git_repo": "https://secp256k1.net/8ball.git",
      "recommended_version": "v1",
      "versions": [
        {
          "binaries": {
            "linux/amd64": "https://8ball.info/8ball.tar.gz",
          },
          "compatible_versions": [
            "v1",
          ],
          "cosmos_sdk_version": "0.46.7",
          "cosmwasm_enabled": true,
          "name": "v1",
          "recommended_version": "v1",
        },
      ],
    },
    "daemon_name": "8ball",
    "explorers": [
      {
        "tx_page": "https://explorer.8ball.info/8ball/tx/\${txHash}",
        "url": "https://explorer.8ball.info/",
      },
      {
        "kind": "ping.pub",
        "tx_page": "https://ping.pub/8ball/tx/\${txHash}",
        "url": "https://ping.pub/8ball",
      },
      {
        "kind": "NodeStake Explorer",
        "tx_page": "https://explorer.nodestake.top/8ball/tx/\${txHash}",
        "url": "https://explorer.nodestake.top/8ball/",
      },
      {
        "kind": "TC Network",
        "tx_page": "https://explorer.tcnetwork.io/8ball/transaction/\${txHash}",
        "url": "https://explorer.tcnetwork.io/8ball",
      },
      {
        "kind": "TC Network",
        "tx_page": "https://explorer.co.id/8ball/tx/\${txHash}",
        "url": "https://explorer.co.id/8ball",
      },
      {
        "kind": "NODEXPLORER",
        "tx_page": "https://explorer.nodexcapital.com/8ball/tx/\${txHash}",
        "url": "https://explorer.nodexcapital.com/8ball",
      },
    ],
    "fees": {
      "fee_tokens": [
        {
          "average_gas_price": 0.025,
          "denom": "uebl",
          "fixed_min_gas_price": 0,
          "high_gas_price": 0.04,
          "low_gas_price": 0,
        },
      ],
    },
    "images": [
      {
        "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/8ball/images/8ball.png",
        "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/8ball/images/8ball.svg",
      },
    ],
    "key_algos": [
      "secp256k1",
    ],
    "logo_URIs": {
      "png": "https://raw.githubusercontent.com/cosmos/chain-registry/master/8ball/images/8ball.png",
      "svg": "https://raw.githubusercontent.com/cosmos/chain-registry/master/8ball/images/8ball.svg",
    },
    "network_type": "mainnet",
    "node_home": "$HOME/.8ball",
    "peers": {
      "persistent_peers": [
        {
          "address": "one.8ball.info:26656",
          "id": "fca96d0a1d7357afb226a49c4c7d9126118c37e9",
          "provider": "8ball",
        },
        {
          "address": "two.8ball.info:26656",
          "id": "aa918e17c8066cd3b031f490f0019c1a95afe7e3",
          "provider": "8ball",
        },
        {
          "address": "rpc.8ball.nodexcapital.com:11056",
          "id": "49778546e7511a1cd6dde65805cd70547c75ce2b",
          "provider": "NodeX Validator",
        },
      ],
      "seeds": [],
    },
    "pretty_name": "8ball",
    "slip44": 118,
    "staking": {
      "staking_tokens": [
        {
          "denom": "uebl",
        },
      ],
    },
    "status": "live",
    "website": "https://8ball.info/",
  },
  "command": "chain",
  "registryDir": "../../packages/chain-registry/chain-registry",
  "tty": true,
}
`;

exports[`Inquirerer prompts user and correctly processes delayed input 2`] = `
[
  "",
  "choose a command
[--command]:
$ 
",
  "> chain
",
  "  asset-lists
",
  "  ibc-data
",
  "  schemas
",
  "",
  "Select a chain:
[--chain]:
$ 
",
  "> 8ball
",
  "  acrechain
",
  "  agoric
",
  "  aioz
",
  "  akash
",
  "  akiro
",
  "  andromeda
",
  "  andromeda1
",
  "  archway
",
  "  arkh
",
  "  assetmantle
",
  "  aura
",
  "  axelar
",
  "  bandchain
",
  "  beezee
",
  "",
]
`;

exports[`Inquirerer prompts user and correctly processes delayed input 3`] = `
[
  "
",
  "
",
]
`;
